hasktorch/src/Torch/Random.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+
+module Torch.Random
+  ( mkGenerator
+  , Generator
+  , randn
+  , randn'
+  , rand
+  , rand'
+  , randint
+  , randint'
+  , normal
+  , normal'
+  ) where
+
+import ATen.Cast
+import ATen.Class (Castable(..))
+import qualified ATen.Const as ATen
+import qualified ATen.Type as ATen
+import Data.Word
+import qualified ATen.Managed.Type.Generator as ATen
+import qualified Torch.Managed.Native as LibTorch
+import Torch.Device
+import Torch.TensorOptions
+import Torch.Tensor
+import Foreign.ForeignPtr
+import System.IO.Unsafe
+import           Control.Concurrent
+import           Control.Concurrent.STM
+import           Control.Monad.IO.Class
+import           Control.Monad.STM
+
+instance Show (TVar (Maybe (ForeignPtr ATen.Generator))) where
+  show _ = "_"
+
+data Generator
+  = Generator
+  { device :: Device
+  , seed :: Word64
+  , generator :: TVar (Maybe (ForeignPtr ATen.Generator))
+  }
+  deriving (Eq,Show)
+
+mkGenerator :: Device -> Word64 -> IO Generator
+mkGenerator device seed =
```

Comment:
This is the private constructor of generator.
If the generator has a pointer to the actual ATen generator, the pointer moves to the next generator.